# 6. Data Manipulation IV - Combining Data - Quiz

## 6.0. Import `polars` and Load Data

In [1]:
import polars as pl

In [3]:
yellow_rides_column_rename_mapping = {
    "VendorID": "vendor_id",
    "RatecodeID": "ratecode_id",
    "PULocationID": "pu_location_id",
    "DOLocationID": "do_location_id",
    "Airport_fee": "airport_fee",
}
march_yellow_rides_df = (
    pl.read_parquet("../data/yellow_tripdata_2024-03.parquet")
    .rename(yellow_rides_column_rename_mapping)
)

In [4]:
zones_df = (
    pl.read_parquet("../data/taxi_zone_lookup.parquet")
    .rename({
        "LocationID": "location_id",
        "Borough": "borough",
        "Zone": "zone",
    })
)

## 6.1 Question 1: Most Common Pickup-Dropoff Zone Pair

Using the `zones_df` combined with the `march_yellow_rides_df`, which `pu_zone` `do_zone` pair had the most rides?

In [5]:
result = (
    march_yellow_rides_df
    .join(
        zones_df.select(pl.all().name.prefix("pu_")),
        on="pu_location_id",
    )
    .join(
        zones_df.with_columns(pl.all().name.prefix("do_")),
        on="do_location_id",
    )
    #### YOUR CODE HERE
)
print(result)

shape: (1, 3)
┌───────────────────────┬───────────────────────┬───────────┐
│ pu_zone               ┆ do_zone               ┆ num_rides │
│ ---                   ┆ ---                   ┆ ---       │
│ str                   ┆ str                   ┆ u32       │
╞═══════════════════════╪═══════════════════════╪═══════════╡
│ Upper East Side South ┆ Upper East Side North ┆ 21477     │
└───────────────────────┴───────────────────────┴───────────┘


1. (Upper East Side South, Upper East Side North)
2. (Erasmus, Astoria)
3. (Upper East Side North, Upper East Side South)
4. (Midtown Center, Upper East Side North)

## 6.2 Question 2: Average Passenger Count for Specific Route

What is the average `passenger_count` for rides that started in the zone "Midtown Center" and ended in the zone "Upper East Side North"?

In [8]:
result = (
    march_yellow_rides_df
    #### YOUR CODE HERE
)
print(result)

shape: (1, 1)
┌─────────────────┐
│ passenger_count │
│ ---             │
│ f64             │
╞═════════════════╡
│ 1.277752        │
└─────────────────┘


1. 13.526644
2. 1.932876
3. 1.277752
4. 21.211721

## 6.3 Question 3: Diagonal Concatenation Result Shape

Take the two toy dataframes below and concatenate them diagonally. What is the shape of the result?

In [41]:
toy_1_df = pl.DataFrame({
    "a": [1, 2, 4],
    "b": [5, 7, 8],
})
toy_2_df = pl.DataFrame({
    "a": [10, 20, 50, 90],
    "b": [50, 3, 99, 3],
    "c": [4, 6, 10, 12]
})
result = (
    #### YOUR CODE HERE
)
print(result)

(7, 3)


1. (6, 2)
2. (7, 3)
3. (8, 3)
4. (7, 2)